File: database_integration/enhance_db/src/bin/17_create_impact_date.rs
/// Create individuals_impact_date table.
///
/// For each individual, compute an impact_date:
///   - birthdate + 35 years
///   - If birth+35 > deathdate, use deathdate
///   - If no birthdate, use deathdate
///   - If neither, skip
///
/// Columns: wikidata_id, name_en, impact_date, impact_date_precision, date_source
use anyhow::Result;
use rusqlite::{params, Connection};
use std::fs;
use std::io::Write;
use std::time::Instant;

const DB_PATH: &str = "data/humans_clean.sqlite3";
const TASK_LOG: &str = "task.log";
const BATCH_SIZE: usize = 100_000;

fn log(msg: &str) {
    let now = chrono_now();
    let line = format!("[{}] {}", now, msg);
    println!("{}", line);
    let mut f = fs::OpenOptions::new()
        .create(true)
        .append(true)
        .open(TASK_LOG)
        .unwrap();
    writeln!(f, "{}", line).unwrap();
}

fn chrono_now() -> String {
    let dur = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap();
    let secs = dur.as_secs();
    let hours = (secs % 86400) / 3600;
    let mins = (secs % 3600) / 60;
    let s = secs % 60;
    format!("{:02}:{:02}:{:02} UTC", hours, mins, s)
}

fn elapsed(start: Instant) -> String {
    let d = start.elapsed();
    let secs = d.as_secs();
    if secs < 60 {
        format!("{}s", secs)
    } else if secs < 3600 {
        format!("{}m {}s", secs / 60, secs % 60)
    } else {
        format!("{}h {}m {}s", secs / 3600, (secs % 3600) / 60, secs % 60)
    }
}

/// Parse a date string like "YYYY-MM-DD" or "-YYYY-MM-DD" (BCE).
/// Returns (year, month, day) or None if invalid.
fn parse_date(s: &str) -> Option<(i64, u32, u32)> {
    let s = s.trim();
    // Skip blank nodes like "_:bn..."
    if s.starts_with("_:") || s.is_empty() {
        return None;
    }

    let (negative, rest) = if let Some(stripped) = s.strip_prefix('-') {
        (true, stripped)
    } else {
        (false, s)
    };

    let parts: Vec<&str> = rest.splitn(3, '-').collect();
    if parts.len() < 3 {
        return None;
    }

    let year: i64 = parts[0].parse().ok()?;
    let month: u32 = parts[1].parse().ok()?;
    let day: u32 = parts[2].parse().ok()?;

    let year = if negative { -year } else { year };
    Some((year, month, day))
}

/// Format a date back to string. Handles negative years (BCE).
fn format_date(year: i64, month: u32, day: u32) -> String {
    if year < 0 {
        format!("-{:04}-{:02}-{:02}", -year, month, day)
    } else {
        format!("{:04}-{:02}-{:02}", year, month, day)
    }
}

/// Add 35 years to a date.
fn add_35_years(year: i64, month: u32, day: u32) -> (i64, u32, u32) {
    let new_year = year + 35;
    // Handle Feb 29 -> Feb 28 for non-leap years
    let new_day = if month == 2 && day == 29 && !is_leap_year(new_year) {
        28
    } else {
        day
    };
    (new_year, month, new_day)
}

fn is_leap_year(year: i64) -> bool {
    if year <= 0 {
        // Proleptic Gregorian: year 0 = 1 BCE, year -1 = 2 BCE
        let y = 1 - year; // convert to positive
        (y % 4 == 0) && ((y % 100 != 0) || (y % 400 == 0))
    } else {
        (year % 4 == 0) && ((year % 100 != 0) || (year % 400 == 0))
    }
}

/// Compare two dates: returns true if (y1,m1,d1) > (y2,m2,d2)
fn date_gt(y1: i64, m1: u32, d1: u32, y2: i64, m2: u32, d2: u32) -> bool {
    if y1 != y2 {
        return y1 > y2;
    }
    if m1 != m2 {
        return m1 > m2;
    }
    d1 > d2
}

fn main() -> Result<()> {
    let total_start = Instant::now();
    log("=== Step 17: Create individuals_impact_date table ===");

    let conn = Connection::open(DB_PATH)?;
    conn.execute_batch(
        "PRAGMA journal_mode=WAL; PRAGMA synchronous=NORMAL; PRAGMA cache_size=-2000000;",
    )?;

    // Drop existing table if present
    conn.execute_batch("DROP TABLE IF EXISTS individuals_impact_date;")?;
    log("[17] Dropped existing individuals_impact_date (if any)");

    // Create the table
    conn.execute_batch(
        "CREATE TABLE individuals_impact_date (
            wikidata_id TEXT PRIMARY KEY,
            name_en TEXT,
            impact_date TEXT,
            impact_date_precision INTEGER,
            date_source TEXT
        );",
    )?;
    log("[17] Created individuals_impact_date table");

    // Count total individuals
    let total: i64 = conn.query_row("SELECT COUNT(*) FROM individuals", [], |r| r.get(0))?;
    log(&format!("[17] Total individuals to process: {}", total));

    // Read all individuals
    let step = Instant::now();
    log("[17] Reading individuals and computing impact dates...");

    let mut stmt = conn.prepare(
        "SELECT wikidata_id, name_en, birthdate, birthdate_precision, deathdate, deathdate_precision
         FROM individuals"
    )?;

    let mut rows_processed: i64 = 0;
    let mut rows_inserted: i64 = 0;
    let mut from_birth: i64 = 0;
    let mut from_death: i64 = 0;
    let mut skipped_no_date: i64 = 0;

    // Collect results in batches to insert
    let rows = stmt.query_map([], |row| {
        Ok((
            row.get::<_, String>(0)?,                // wikidata_id
            row.get::<_, Option<String>>(1)?,         // name_en
            row.get::<_, Option<String>>(2)?,         // birthdate
            row.get::<_, Option<i64>>(3)?,            // birthdate_precision
            row.get::<_, Option<String>>(4)?,         // deathdate
            row.get::<_, Option<i64>>(5)?,            // deathdate_precision
        ))
    })?;

    // Begin transaction
    conn.execute_batch("BEGIN TRANSACTION;")?;

    {
        let mut insert_stmt = conn.prepare(
            "INSERT INTO individuals_impact_date (wikidata_id, name_en, impact_date, impact_date_precision, date_source)
             VALUES (?1, ?2, ?3, ?4, ?5)"
        )?;

        for row_result in rows {
            let (wikidata_id, name_en, birthdate, birth_precision, deathdate, death_precision) =
                row_result?;

            rows_processed += 1;

            let birth_parsed = birthdate.as_deref().and_then(parse_date);
            let death_parsed = deathdate.as_deref().and_then(parse_date);

            let result: Option<(String, Option<i64>, &str)> = match (birth_parsed, death_parsed) {
                (Some((by, bm, bd)), Some((dy, dm, dd))) => {
                    // Have both dates
                    let (iy, im, id) = add_35_years(by, bm, bd);
                    if date_gt(iy, im, id, dy, dm, dd) {
                        // birth+35 > death, use death date
                        Some((format_date(dy, dm, dd), death_precision, "deathdate"))
                    } else {
                        // Use birth+35
                        Some((format_date(iy, im, id), birth_precision, "birthdate"))
                    }
                }
                (Some((by, bm, bd)), None) => {
                    // Only birthdate
                    let (iy, im, id) = add_35_years(by, bm, bd);
                    Some((format_date(iy, im, id), birth_precision, "birthdate"))
                }
                (None, Some((dy, dm, dd))) => {
                    // Only deathdate
                    Some((format_date(dy, dm, dd), death_precision, "deathdate"))
                }
                (None, None) => {
                    // No dates
                    skipped_no_date += 1;
                    None
                }
            };

            if let Some((impact_date, precision, source)) = result {
                match source {
                    "birthdate" => from_birth += 1,
                    _ => from_death += 1,
                }
                insert_stmt.execute(params![
                    wikidata_id,
                    name_en,
                    impact_date,
                    precision,
                    source
                ])?;
                rows_inserted += 1;
            }

            if rows_processed % 1_000_000 == 0 {
                log(&format!(
                    "[17]   Processed {}/{} ({:.1}%) - inserted: {} ({}) ",
                    rows_processed,
                    total,
                    (rows_processed as f64 / total as f64) * 100.0,
                    rows_inserted,
                    elapsed(step)
                ));
            }
        }
    }

    conn.execute_batch("COMMIT;")?;
    log(&format!(
        "[17] All rows processed ({}) in {}",
        rows_processed,
        elapsed(step)
    ));

    // Create index
    let idx_start = Instant::now();
    log("[17] Creating indexes...");
    conn.execute_batch(
        "CREATE INDEX IF NOT EXISTS idx_impact_date ON individuals_impact_date(impact_date);
         CREATE INDEX IF NOT EXISTS idx_impact_wid ON individuals_impact_date(wikidata_id);"
    )?;
    log(&format!("[17] Indexes created ({})", elapsed(idx_start)));

    // Summary
    log(&format!("[17] === Summary ==="));
    log(&format!("[17]   Total processed: {}", rows_processed));
    log(&format!("[17]   Inserted: {}", rows_inserted));
    log(&format!("[17]   From birthdate+35: {}", from_birth));
    log(&format!("[17]   From deathdate: {}", from_death));
    log(&format!("[17]   Skipped (no dates): {}", skipped_no_date));

    // Verify
    let verify: i64 = conn.query_row(
        "SELECT COUNT(*) FROM individuals_impact_date",
        [],
        |r| r.get(0),
    )?;
    log(&format!("[17]   Verified row count: {}", verify));

    // Sample output
    log("[17] Sample rows:");
    let mut sample = conn.prepare(
        "SELECT wikidata_id, name_en, impact_date, impact_date_precision, date_source
         FROM individuals_impact_date LIMIT 5"
    )?;
    let samples: Vec<(String, Option<String>, String, Option<i64>, String)> = sample
        .query_map([], |r| {
            Ok((
                r.get(0)?,
                r.get(1)?,
                r.get(2)?,
                r.get(3)?,
                r.get(4)?,
            ))
        })?
        .filter_map(|r| r.ok())
        .collect();
    for (wid, name, date, prec, src) in &samples {
        log(&format!(
            "[17]   {} | {} | {} | prec={:?} | src={}",
            wid,
            name.as_deref().unwrap_or("NULL"),
            date,
            prec,
            src
        ));
    }

    log(&format!(
        "=== Step 17 complete (total: {}) ===",
        elapsed(total_start)
    ));
    Ok(())
}

File: scripts/database_integration_scripts_V2/legacy/17_create_impact_date.py
"""17 — Build the `individuals_impact_date` table.

Mirrors `enhance_db/src/bin/17_create_impact_date.rs`.

For each individual:
  * `min(birthdate + 35 years, deathdate)` if both dates are present
  * `birthdate + 35 years` if only birthdate is present
  * `deathdate` if only deathdate is present
  * skipped if neither

  Inputs : individuals (wikidata_id, name_en, birthdate*, deathdate*)
  Output : individuals_impact_date (wikidata_id PK, name_en, impact_date,
           impact_date_precision, date_source)

Usage
-----
    python3 17_create_impact_date.py
    python3 17_create_impact_date.py --full
"""

from __future__ import annotations

import sqlite3
import tempfile
from pathlib import Path

from common import (
    DB_PATH,
    insert_rows,
    log,
    open_db,
    parse_run_mode,
    transaction,
)


def _is_leap(year: int) -> bool:
    y = (1 - year) if year <= 0 else year
    return (y % 4 == 0) and ((y % 100 != 0) or (y % 400 == 0))


def _parse_date(s: str | None):
    if not s:
        return None
    s = s.strip()
    if not s or s.startswith("_:"):
        return None
    negative = s.startswith("-")
    rest = s[1:] if negative else s
    parts = rest.split("-", 2)
    if len(parts) < 3:
        return None
    try:
        y = int(parts[0])
        m = int(parts[1])
        d = int(parts[2].split("T", 1)[0])
    except ValueError:
        return None
    return (-y if negative else y, m, d)


def _add_35_years(y: int, m: int, d: int) -> tuple[int, int, int]:
    new_y = y + 35
    new_d = 28 if (m == 2 and d == 29 and not _is_leap(new_y)) else d
    return (new_y, m, new_d)


def _format_date(y: int, m: int, d: int) -> str:
    if y < 0:
        return f"-{-y:04d}-{m:02d}-{d:02d}"
    return f"{y:04d}-{m:02d}-{d:02d}"


def _date_gt(a, b) -> bool:
    return a > b


def run(conn: sqlite3.Connection) -> None:
    log("[DB] 17: Creating individuals_impact_date...")

    conn.execute("DROP TABLE IF EXISTS individuals_impact_date")
    conn.execute(
        """
        CREATE TABLE individuals_impact_date (
            wikidata_id TEXT PRIMARY KEY,
            name_en TEXT,
            impact_date TEXT,
            impact_date_precision INTEGER,
            date_source TEXT
        )
        """
    )

    total = conn.execute("SELECT COUNT(*) FROM individuals").fetchone()[0]
    log(f"[17] Total individuals: {total}")

    cur = conn.execute(
        "SELECT wikidata_id, name_en, birthdate, birthdate_precision, "
        "deathdate, deathdate_precision FROM individuals"
    )

    try:
        from tqdm import tqdm
        iterator = tqdm(cur, total=total, desc="impact_date", unit="row")
    except ImportError:
        iterator = cur

    rows_processed = 0
    rows_inserted = 0
    from_birth = 0
    from_death = 0
    skipped_no_date = 0

    insert_sql = (
        "INSERT INTO individuals_impact_date "
        "(wikidata_id, name_en, impact_date, impact_date_precision, date_source) "
        "VALUES (?, ?, ?, ?, ?)"
    )
    BATCH = 100_000
    buf: list[tuple] = []

    with transaction(conn):
        ins = conn.cursor()
        for wid, name_en, birthdate, birth_prec, deathdate, death_prec in iterator:
            rows_processed += 1
            bp = _parse_date(birthdate)
            dp = _parse_date(deathdate)

            if bp is not None and dp is not None:
                impact = _add_35_years(*bp)
                if _date_gt(impact, dp):
                    buf.append((wid, name_en, _format_date(*dp), death_prec, "deathdate"))
                    from_death += 1
                else:
                    buf.append((wid, name_en, _format_date(*impact), birth_prec, "birthdate"))
                    from_birth += 1
                rows_inserted += 1
            elif bp is not None:
                impact = _add_35_years(*bp)
                buf.append((wid, name_en, _format_date(*impact), birth_prec, "birthdate"))
                from_birth += 1
                rows_inserted += 1
            elif dp is not None:
                buf.append((wid, name_en, _format_date(*dp), death_prec, "deathdate"))
                from_death += 1
                rows_inserted += 1
            else:
                skipped_no_date += 1

            if len(buf) >= BATCH:
                ins.executemany(insert_sql, buf)
                buf.clear()
        if buf:
            ins.executemany(insert_sql, buf)
            buf.clear()

    conn.execute(
        "CREATE INDEX IF NOT EXISTS idx_impact_date ON individuals_impact_date(impact_date)"
    )
    conn.execute(
        "CREATE INDEX IF NOT EXISTS idx_impact_wid ON individuals_impact_date(wikidata_id)"
    )
    conn.commit()

    log("[17] === Summary ===")
    log(f"[17]   Total processed: {rows_processed}")
    log(f"[17]   Inserted: {rows_inserted}")
    log(f"[17]   From birthdate+35: {from_birth}")
    log(f"[17]   From deathdate: {from_death}")
    log(f"[17]   Skipped (no dates): {skipped_no_date}")


def _sample_main() -> None:
    with tempfile.TemporaryDirectory() as tmp:
        db = Path(tmp) / "sample.sqlite3"
        with sqlite3.connect(db) as seed:
            seed.execute(
                """
                CREATE TABLE individuals (
                    wikidata_id TEXT PRIMARY KEY,
                    name_en TEXT,
                    birthdate TEXT,
                    birthdate_precision INTEGER,
                    deathdate TEXT,
                    deathdate_precision INTEGER
                )
                """
            )
            insert_rows(seed, "individuals", [
                {"wikidata_id": "Q1", "name_en": "Long Lived", "birthdate": "1800-01-01",
                 "birthdate_precision": 11, "deathdate": "1900-01-01", "deathdate_precision": 11},
                {"wikidata_id": "Q2", "name_en": "Short Lived", "birthdate": "1850-06-15",
                 "birthdate_precision": 11, "deathdate": "1860-01-01", "deathdate_precision": 11},
                {"wikidata_id": "Q3", "name_en": "Only Birth", "birthdate": "1900-03-10",
                 "birthdate_precision": 11, "deathdate": None, "deathdate_precision": None},
                {"wikidata_id": "Q4", "name_en": "Only Death", "birthdate": None,
                 "birthdate_precision": None, "deathdate": "0500-12-25", "deathdate_precision": 11},
                {"wikidata_id": "Q5", "name_en": "BCE", "birthdate": "-0050-03-15",
                 "birthdate_precision": 9, "deathdate": "0014-08-19", "deathdate_precision": 11},
                {"wikidata_id": "Q6", "name_en": "Empty", "birthdate": None,
                 "birthdate_precision": None, "deathdate": None, "deathdate_precision": None},
            ])

        with open_db(db) as conn:
            run(conn)
            for row in conn.execute(
                "SELECT wikidata_id, name_en, impact_date, impact_date_precision, date_source "
                "FROM individuals_impact_date ORDER BY wikidata_id"
            ):
                log(f"  {row}")


if __name__ == "__main__":
    if parse_run_mode() == "full":
        with open_db(DB_PATH) as conn:
            run(conn)
    else:
        _sample_main()
